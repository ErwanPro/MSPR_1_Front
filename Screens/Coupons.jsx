import React from "react";
import { FlatList, RefreshControl } from "react-native";
import { connect } from "react-redux";
import Coupon from "../Components/Coupon";

export class Coupons extends React.Component {
	/**
	 * 
	 * @param {*} props 
	 */
	constructor(props) {
		super(props);

		this.state = {
			refreshing: false,
		};
	}

	/**
	 * 
	 * @returns Liste de l'ensemble des coupons scannés
	 */
	render() {
		return (
			<FlatList
				style={{ backgroundColor: "white" }}
				data={this.props.coupons}
				renderItem={({ item }) => <Coupon {...item} key={item._id} />}
				refreshControl={<RefreshControl refreshing={this.state.refreshing} onRefresh={this.onRefresh} />}
			/>
		);
	}
}

/**
 * 
 * @param {*} state 
 * @returns Coupons présents dans le store global
 */
const mapStateToProps = state => {
	return {
		coupons: state.coupons,
	};
};

export default connect(mapStateToProps)(Coupons);
