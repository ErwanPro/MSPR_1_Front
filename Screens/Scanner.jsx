import React from "react";
import { View, Text, StyleSheet } from "react-native";
import { connect } from "react-redux";


import { BarCodeScanner } from "expo-barcode-scanner";

import { getCouponById } from "../API/Coupon.js";

class Scanner extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			scanned: false,
			flash: false,
			cameraPermission: false,
		};
	}

	async componentDidMount() {
		const { status } = await BarCodeScanner.requestPermissionsAsync();
		this.setState({ cameraPermission: status === "granted" });

		const unsubscribeFocus = this.props.navigation.addListener('focus', () => {
			this.setState({scanned: false});
		});
		
		this.unsubscribeFocus = unsubscribeFocus;
		// const unsubscribeBlur = this.props.navigation.addListener('blur', () => {

		// });
	}

	componentWillUnmount() {
		this.unsubscribeFocus();
	}

	handleBarCodeScanned = async event => {
		const idCoupon = event.data.replace("gs://", "");
		try {
			const coupon = await getCouponById(idCoupon);
			coupon.id = coupon._id;
			delete coupon._id;
			delete coupon.__v;
			const action = { type: "ADD_COUPON", ...coupon };
			console.log(action);
			this.props.dispatch(action);
			// this.ScanCallBack();
			this.setState({ scanned: true });
			this.props.navigation.jumpTo("List");
		} catch (e) {
			console.error(e);
		}
	};

	ScanCallBack = () => {
		this.setState({ scanned: true }, () => {
			setTimeout(() => {
				this.setState({ scanned: false });
			}, 2000);
		});
	}

	render() {
		if (this.state.cameraPermission) {
			return (
				<View>
					<BarCodeScanner onBarCodeScanned={!this.state.scanned && this.handleBarCodeScanned} style={styles.barcode} />
				</View>
			);
		} else {
			return (
				<View>
					<Text>L'application n'a pas access a la camera</Text>
				</View>
			);
		}
	}
}

const styles = StyleSheet.create({
	barcode: {
		height: 800,
		width: 400,
	},
});

export default connect()(Scanner);
