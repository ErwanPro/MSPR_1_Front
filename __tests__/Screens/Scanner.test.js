import React from "react";
import renderer from 'react-test-renderer';
import { Scanner } from '../../Screens/Scanner';

it(`renders correctly`, () => {
  const tree = renderer.create(<Scanner />).toJSON();
  expect(tree).toMatchSnapshot();
});