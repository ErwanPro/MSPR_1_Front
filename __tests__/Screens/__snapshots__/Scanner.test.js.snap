// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <Text>
    L'application n'a pas access a la camera
  </Text>
</View>
`;
